### PR TITLE
VSTS support

### PR DESCRIPTION
**Pull request by sgmunn**
This introduces the notion that there is more than one provider of issues, namely VSTS but allows for additional providers such as GitHub issues.

Commit parsing and validation have been separated into 2 distinct processes. This allows us to test parsing separate from making network calls to bugzilla or vsts. It also allows us to request a list of issues to be discovered from commits without any network calls.

Parsing Confidence has been altered to allow us to separate issues that have a high confidence (ie when the commit references the full url of an issue) compared to vague bug references (eg "bug 1234").

New command options

- x, export-bugs: Allows us to export the results of the parsing and validation steps to xml (with --output-file). I need this to be able to use this upload this as an artefact when run from a VSTS build

- commit=: Allows a simpler way of processing just one commit. This will be when run from VSTS build for a given commit to generate a report on what issue(s) are fixed in that commit

- vsts=: enables or disables VSTS validation like bugzilla=private/disable

- ignore-vsts: disables VSTS parsing and validation completely

- ignore-bugzilla: disables bugzilla parsing and validation completely

Submodule processing has been tweaked to handle single commit ranges. There is an open TODO around whether to exclude the oldest commit in the submodule range or not. I believe that it should be.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [submodule "clio-test-data"]
 	path = clio-test-data
-	url = git@github.com:chamons/clio-test-data.git
```
Comment (chamons) [inline .gitmodules:3]: Seems unnecessary diff. Revert?

File: src/Providers/HttpService.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+				req.KeepAlive = false;
+				req.Expect = "";
+				req.AllowWriteStreamBuffering = true;
+				req.Method = method ?? "POST";
+				req.Headers.Add ("Content-Encoding", "utf-8");
+				if (accept != null)
+					req.Accept = accept;
+				if (contentType != null)
+					req.ContentType = contentType;
+				//				if (user != null)
```
Comment (chamons) [inline src/Providers/HttpService.cs]: Commented out code?

File: src/Providers/HttpService.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+				WebResponse resp = await req.GetResponseAsync ();
+				HttpWebResponse httpResp = resp as HttpWebResponse;
+				if (expectedStatus != null && httpResp != null && (int)httpResp.StatusCode != expectedStatus.Value)
+					throw new WebException (string.Format ("Expected status {0}, but got {1}: {2}", expectedStatus.Value, httpResp.StatusCode, httpResp.StatusDescription));
+
+				return resp.GetResponseStream ();
+			}
+			catch (WebException)
+			{
+				//using (StreamReader sr = new StreamReader (ex.Response.GetResponseStream ()))
```
Comment (chamons) [inline src/Providers/HttpService.cs]: Same

File: src/Providers/HttpService.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+				//	Console.WriteLine ("resp: " + sr.ReadToEnd ());
+				throw;
+			}
+		}
+
+		public async Task<Stream> GetAsStreamAsync (string url, string user = null, string pwd = null)
+		{
+			Console.WriteLine ("GET " + url);
```
Comment (chamons) [inline src/Providers/HttpService.cs]: Debug code?

File: src/Providers/VisualStudioService.cs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+				}
+				catch (WebException ex)
+				{
+					ParseError (ex);
+					throw;
+				}
+			}
+			catch
+			{
+				// TODO: log errors somehow
```
Comment (chamons) [inline src/Providers/VisualStudioService.cs]: Should this scream in explain? Or is it safe to ignore.

**Review by chamons — APPROVED** [review]
A few minor nits to clean up but looks wonderful!

File: .gitmodules (hunk 1 above)
Comment (sgmunn) [reply .gitmodules:3]: this is needed to build on VSTS CI. the ssh link doesn't work.

File: src/Providers/HttpService.cs (hunk 2 above)
Comment (sgmunn) [reply src/Providers/HttpService.cs]: that's what you get when you yank code from someplace else. will tidy up.

File: src/Providers/VisualStudioService.cs (hunk 5 above)
Comment (sgmunn) [reply src/Providers/VisualStudioService.cs]: yeah, forgot about Explain for that. I just didn't want to crash the entire run for weird / network issues.

**Comment by sgmunn** [comment]
@chamons I'll tidy these issues up in a little while.

**Comment by sgmunn** [comment]
@chamons cleaned up nits.

**Review by chamons — APPROVED** [review]
